src/masternode-sync.cpp

```diff
@@ -53,24 +53,19 @@ std::string CMasternodeSync::GetAssetName()
 
 void CMasternodeSync::SwitchToNextAsset(CConnman& connman)
 {
-    CDeterministicMNList mnList;
     switch(nCurrentAsset)
     {
         case(MASTERNODE_SYNC_FAILED):
             throw std::runtime_error("Can't switch to next asset from failed, should use Reset() first!");
             break;
         case(MASTERNODE_SYNC_INITIAL):
-            // try to update UI with the current MN list/count (can fail since we might be still syncing)
-            mnList = deterministicMNManager->GetListAtChainTip();
-            uiInterface.NotifyMasternodeListChanged(mnList);
             nCurrentAsset = MASTERNODE_SYNC_WAITING;
             LogPrintf("CMasternodeSync::SwitchToNextAsset -- Starting %s\n", GetAssetName());
             break;
         case(MASTERNODE_SYNC_WAITING):
             LogPrintf("CMasternodeSync::SwitchToNextAsset -- Completed %s in %llds\n", GetAssetName(), GetTime() - nTimeAssetSyncStarted);
             // blockchain is synced, update UI with the correct MN list/count
-            mnList = deterministicMNManager->GetListAtChainTip();
-            uiInterface.NotifyMasternodeListChanged(mnList);
+            uiInterface.NotifyMasternodeListChanged(deterministicMNManager->GetListAtChainTip());
```

Comment:
Isn't this unneeded due to `CDeterministicMNManager` calling NotifyMasternodeListChanged?

Reply:
This one is still needed because we do not process every signal while we are still syncing, so we could skip some very recent ones and UI would not be quite up to date without this.

Reply:
Hmm ok, understand. However, while looking into the code I found another issue and fixing it might be related to the problem fixed here.

The check in https://github.com/dashpay/dash/blob/b077507c0686d012b866babe0740ded97a51a86c/src/evo/deterministicmns.cpp#L484
results in no calls to `NotifyMasternodeListChanged` when the meta data of MNs changes, e.g. keys, IPs, last paid block, ...
Actually, such a changes happen for every single block, simply due to payment of MNs in every block. To fix this, we could change the surrounding if to use `diff.HasChanges()`, however, this is kind of bogus and misleading as we know that the list "changes" every block. Better solution is probably to just remove the surrounding if and call `NotifyMasternodeListChanged` for every block.

This in turn, makes the special handling in `src/masternode-sync.cpp` kind of unnecessary, as the next block after entering `MASTERNODE_SYNC_WAITING` will invoke `MasternodeListChanged` (from clientmodel.cpp) with `initialSync == false`, resulting in an up-to-date list in the client model.

Reply:
Hmmm... but if this happens every block, do we need `NotifyMasternodeListChanged` at all?

Reply:
Hmm, that's a good point as well. Just reacting to `NotifyBlockTip` should be enough then

Reply:
Pls see 4b4778f

Reply:
On another thought, 4b4778f was doing way too much probably. Force-pushed lighter (UI only) version - 0675702.

Reply:
I assume you want to put the other changes from https://github.com/dashpay/dash/commit/4b4778f920943fb2607d56dbb7c5952a796c33e3 into a new PR then?

Reply:
Maybe :) I mean, there are like two different meanings to "masternode list changed" - one is visual/meta stuff (on each block) and the other one is when MN is actually in/out of the list. Maybe we can still handle the later on each block too, I'm not sure anymore tbh 🙈 But in any case we shouldn't be pushing it in "MN list UI vs data" PR imo.